### ISSUE TITLE
Suggestion - End of task indicates end of compilation

### ISSUE THREAD

**Issue by gabrielschnegelberger** [ivogabe/gulp-typescript#248]
I experienced some heavy problems with gulp-typescript and browserify. (Or any other task that requires the output of gulp-typescript)

Scenario - Two tasks: The first one compiles the Typescript (with CommonJS Module Resolution) in first glance; and then a second one - which uses browserify to bundle everything up!

To ensure that everything is working correctly I build up the dependencys - but browserify didn't find anything at all (especially if the project gets larger and compilication time increases)

I noticed that the **ts** returns immediately and leaves the typescript compiler in the background working alone. The workaround was to use the callback-Style for grunt-tasks and call it on the **end** event. 
It was additional necessary to add an timer - because it seems that the system requires some time to write the files to the disk. (must be estimated :-/ )

It would be a great thing if there is for example a additional parameter which determines if you want to wait until the compilation is done (100%) and the files are written to the disk - and can be then used surely by the next task!

I'm sure more developers outside use browserify or webpack ;-)

**My workaround so far**
```javascript
function tsCompile(src, dst, finishCallback) {
    var outputDir = "./ts-tmp";
    var tsPipeline = gulp.src( [ src ])
        .pipe(ts({
            noImplicitAny: true,
            noEmitOnError: true,
            target: "ES5",
            removeComments: true,
            module: "commonjs",
            moduleResolution: "classic",
            isolatedModules: true
        }) );

    tsPipeline.pipe( gulp.dest(outputDir + dst) );

    tsPipeline.on("end", function() {
        console.log("Typescript End Compiling " + src);

        // Timeout Time = time which system needs to write down the files
        // values below 1000ms seem for full compiliation too less (with modern Intel i5 CPU)
        setTimeout( function() {
            finishCallback();
        }, 1000);
    });
}

// Use task with callback = means that task is synchronous! 
// Would be the same like "return tsCompile(...);" 
gulp.task("moduleTsCompile", ["tsCompile"], function(finishCallback) {
    tsCompile("modules/**/*.ts", "/modules", finishCallback);
});
```

**Comment by ivogabe** [comment]
If you return the stream, the task will finish when the compilation is finished. Example:
```javascript
gulp.task("scripts", function() {
    return gulp.src("modules/**/*.ts") // Notice the return statement
        .pipe(ts({ module: "commonjs" }))
        .pipe(gulp.dest("dist"));
});
```

**Comment by ivogabe** [comment]
Or, if you want to have a simple gulpfile, you could also use [small](https://www.npmjs.com/package/small) instead of Browserify:
```javascript
gulp.task("scripts", function() {
    return gulp.src("modules/**/*.ts")
        .pipe(ts({ module: "commonjs" }))
        .pipe(small())
        .pipe(gulp.dest("dist"));
});
```

**Comment by gabrielschnegelberger** [comment]
You are right - if I return the entire pipeline it really ends when it's finished ... (I had changed my code over hunderd times already ...)

Still browserify fails every 10th time - reason is not obvious! (Even with extra timeout) 
So I think it's a browserify problem - because it generates a blank file; and doesn't output any error at all! (At first I thought that it might not get the javascript files "quick" enough)

I tried your small-Plugin which seems very interested - but it showed some error that it can't access the **standalone** Property (of undefined?) ... 

Because I have no index.js File - I just called: .`pipe(small('/modules/app.module.js'))` - which showed the error the first time. I then tried to set the options according to the error: 
`pipe(   small('/modules/app.module.js', { standalone: "..."  ]  )   )` - which showed the exactly same error 


But to be honest - I want to stay behind Browserify at the moment ... even if it works not every time :-)

Thanks anyway!
